src/web/lib/components/StorageSpaceInformation/index.js

```diff
@@ -0,0 +1,57 @@
+import React from "react";
+import { connect } from "react-redux";
+import { compose } from "redux";
+import "./index.scss";
+
+export const STORAGE_ERROR_MESSAGE =
+  "Sorry! You cannot add this image/theme as it will put you over your storage capacity. Please delete some saved images/themes to make space.";
+
+export const STORAGE_WARNING_MESSAGE =
+  "You have almost reached your storage limit! You may need to delete some saved images/themes to make space.";
+
+export const STORAGE_ERROR_MESSAGE_DURATION = 5000;
+
+export const localStorageSpace = () => {
+  let data = "";
+  for (let key in window.localStorage) {
+    if (window.localStorage.hasOwnProperty(key)) {
```

Comment:
The size of the keys should also be included in the count.
```
let size = 0;
for (let [key, value] of Object.entries(localStorage)) {
  size += key.length + value.length;
}
```

The limit can be configured through the `dom.storage.default_quota` pref, defaulting to 5120 kB (= `5 x 1024`), i.e. 5242880 "bytes" (the actual implementation doesn't count the number of bytes, but the length of the string, which could be way more if the string contains multi-byte characters).

This means that the limit below (10 MB) overestimates the amount of available storage. The actual limit is `5 * 1024 * 1024`.

Firefox Color doesn't efficiently save images either: it base64-encodes images, which increases the storage size of images by 33%. This means that the total size of images is at most 5  / 1.33 = 3.76 MB. Minus other metadata (themes and such), there is not that much storage left :/

Reply:
thank for the feedback @Rob--W :) I can adjust the calculation as suggested. 
yes, we realized the images were the main culprit (we talked about possibly compressing these but I did not include that with this PR). 

I'll take a look this evening:)

Reply:
ok so I've modified the logic to include the key length as well. I have also adjusted this to be `[number]MB out of 5.242MB` (instead "..out of 10MB"):

``` let size = 0;
  for (let [key, value] of Object.entries(localStorage)) {
    size += key.length + value.length;
  }

  const sizeMB = size / 1000000;
  return parseFloat(sizeMB.toFixed(3));
```

Originally I was hitting the ~10MB when I was nearing the storage limit but w/ the updated calculation it's about this 5.24MB. Please let me know if I have this right. If so I'll push this up:)

Reply:
Your chosen limit looks right. Do note that the limit is dependent on the browser settings, and that most users will have the default of 5 MiB. The limit should be informative, but not prevent the user from adding more images (and updating and exporting themes) if they can (to test that, visit `about:config` and set `dom.storage.default_quota` to a higher value, or simply change your application logic to "enforce" a lower-than-maximum limit). Similarly, if the actual limit is lower, then the application should still handle that gracefully.

Reply:
@caitmuenster, I know we talked about this but it does look like the default limit is ~5.24MB after adjusting the calculation as noted above. So I've updated the message to look something like the following ~ :

![](https://monosnap.com/image/KqVU6RMKjRcaYEQlOjcAzTLGhHY9vt.png)

Also, as far as I know there is no way to find out if the user has changed the default so I am not sure if we want to make an additional note about this or this will be okay as is given most users will probably have the default. 

@Rob--W,
Maybe I am misunderstanding what you mean here:

>The limit should be informative, but not prevent the user from adding more images..

If the storage is full, it won't allow additional images to be added so in this case it would prevent the user from adding more images unless these are cleared out.

Regardless, the user will see a message ~ "Sorry! This cannot be added because you are over your storage limit...." if it will put them over but again the "..out of [storageLimit]" message <= might not be completely accurate if they're not using the default. 

Also, I noticed if the image is not in local storage (it can be removed by the X), it won't be included in the export (cc @caitmuenster). They would need to export it before removing it if they want to include it.